### ISSUE TITLE
Pretty Printers for plain-old data types

### ISSUE THREAD

**Issue by scopatz** [scopatz/xonsh#294]
Maybe look at what IPython does.

**Comment by melund** [comment]
It would be a good idea to have it build into xonsh. 

 I have added this to my xonshrc file as a temp workaround. 

``` python
import sys
from IPython.lib import pretty as _pretty 

__pprint_orginal_hook = sys.displayhook
def __pprint_hook(value):
    if value != None:
        __builtins__['_'] = value
        _pretty.pprint(value)

aliases['pprint_on'] = lambda: setattr(sys, 'displayhook', __pprint_hook)
aliases['pprint_off'] =  lambda: setattr(sys, 'displayhook', __pprint_orginal_hook)
# Activate pretty printing
aliases['pprint_on']()

```

**Comment by scopatz** [comment]
Thanks for the work around , @melund.  In the last line, though, you can just write `pprint_on`.

**Comment by melund** [comment]
You are right @scopatz .... The lambda functions should be different though. 

``` python        
aliases['pprint_on'] = lambda args, stdin=None: setattr(sys, 'displayhook', __pprint_hook)
aliases['pprint_off'] =  lambda args, stdin=None: setattr(sys, 'displayhook', __pprint_orginal_hook)
pprint_on
``` 

How would you prefer that something like this was built into xonsh?

/Morten

**Comment by scopatz** [comment]
I'd prefer that IPython's pretty functionality be copied into xonsh (with appropriate copyright notices, of course, like else where in the code base.) Then, for `displayhook()` to be set to this automatically all of the time. I can't think of a reason why it would be desirable to turn this off and on, but I could be wrong on this point.  

I am about to hop on a plane, so probably won't get to this tonight.  A PR would be very welcome :)

**Comment by melund** [comment]
That seems simple enough. In which source file would you place the code that sets the display hook? I took a look at the IPython.lib.pretty module. It has a tiny internal dependency that handles py2/3 issues. That is not needed as long as xonsh is only python 3. 

``` python
from IPython.utils.py3compat import PY3, cast_unicode
from IPython.utils.encoding import get_stream_enc
...
...
if PY3:
    CUnicodeIO = StringIO
else:
    class CUnicodeIO(StringIO):
        """StringIO that casts str to unicode on Python 2"""
        def write(self, text):
            return super(CUnicodeIO, self).write(
                cast_unicode(text, encoding=get_stream_enc(sys.stdout)))
```

**Comment by scopatz** [comment]
> In which source file would you place the code that sets the display hook?

I'd probably do this in main.py

> It has a tiny internal dependency that handles py2/3 issues. 

I wouldn't worry about bringing this over.  As far as I know, there are no efforts to currently backport xonsh to py2k.

**Comment by melund** [comment]
Ok. I made a PR that ports the pretty print functionality from IPython. 
https://github.com/scopatz/xonsh/pull/303
How do we give proper credits.

**Comment by scopatz** [comment]
Since IPython is BSD, we just need to place the copyright notice in the code.  I commented in #303 where and how this should be done.

**Comment by scopatz** [comment]
closed via #303
